src/nodes/mod.rs

```diff
@@ -0,0 +1,56 @@
+// Copyright 2018 Google Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License"); you may not
+// use this file except in compliance with the License.  You may obtain a copy
+// of the License at:
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+// WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.  See the
+// License for the specific language governing permissions and limitations
+// under the License.
+
+extern crate fuse;
+
+use std::ffi::OsStr;
+use std::io;
+use std::sync::Arc;
+
+mod dir;
+pub use self::dir::Dir;
+
+/// Abstract representation of a file system node.
+///
+/// Due to the way nodes and node operations are represented in the kernel, this trait exposes a
+/// collection of methods that do not all make sense for all possible node types: some methods will
+/// only make sense for directories and others will only make sense for regular files, for example.
+/// These conflicting methods come with a default implementation that panics.
```

Comment:
Dislike panics. Can we avoid the panics with type magic? Or at least have the default impl return some kind of reasonable error (with log if possible).

Reply:
The reason I went for the panics (with this design) is because node implementations will need to be fully specified. For example: the `Symlink` node has to implement all unimplemented methods (like `inode`) and must only override the `readlink` method. If we end up calling any of the other methods with defaults (like `lookup`), that's a big problem because it makes no sense to do so on a symlink node. I'd rather know these things with a crash than to continue and cause potential corruption. That's why I don't think "sane" default implementations are a good thing here: they are bound to hide subtle design problems.

But I agree with you: they are ugly. I'd very much prefer some type magic to only call things when they are defined, but to me that seems 1. hard due to the dynamic nature of nodes, and 2. maybe slower? Inspecting types at run time sounds costly, at least in other languages. I'm afraid I don't know enough Rust to propose a better solution though. Any suggestions?

Reply:
I don't have enough experience with rust types to tell you if type magic will do here. You might want to inquire on the discourse server to see if more experienced people have some ideas.

Reply:
Why can't we just return an error to fuse? Then the error can be propagated back to the user without crashing the process. For example `readdir` can return `ENOTDIR` by default.